core/src/main/java/org/apache/calcite/rel/RelNode.java

```diff
@@ -335,6 +335,26 @@ default String getDigest() {
   @API(since = "1.24", status = API.Status.INTERNAL)
   void recomputeDigest();
 
+  /**
+   * Equality check for RelNode digest.
+   *
+   * <p>By default this method collects digest attributes from
+   * explain terms, then compares each attribute pair.</p>
+   *
+   * @return Whether the 2 RelNodes are equivalent or have the same digest.
+   * @see #digestHash()
+   */
+  @API(since = "1.25", status = API.Status.EXPERIMENTAL)
+  boolean digestEquals(Object obj);
+
+  /**
+   * Compute hash code for RelNode digest.
+   *
+   * @see #digestEquals(Object)
+   */
+  @API(since = "1.25", status = API.Status.EXPERIMENTAL)
+  int digestHash();
```

Comment:
Does it make the code clearer by extracting the two methods to a separate interface, as the RelNode interface is already large enough?

Reply:
I don't think so. It should be originally part of RelNode, and actually should replace hashCode() (if it isn't final), IMO.

Reply:
I see. Thanks for your clarification.
Do we intend to support this for other tree-like data structures (e.g. RexNode) ?

Reply:
RexNode already supports deep equals.